### PR TITLE
[YANG] Add support for identity and identityref

### PR DESCRIPTION
**Pull request by dpino**
Fixes https://github.com/Igalia/snabb/issues/798

### CODE REVIEW

File: src/lib/yang/schema.lua [hunk 1]
```diff
@@ -502,8 +508,10 @@ declare_initializer(
    'error-message', 'error-app-tag', 'max-value', 'key', 'unique', 'when',
    'deviation', 'deviate')
 declare_initializer(
-   init_natural, 'yang-version', 'fraction-digits', 'position',
-   'min-elements', 'max-elements')
+   init_natural, 'fraction-digits', 'position', 'min-elements', 'max-elements')
+declare_initializer(
```
Comment (wingo) [inline src/lib/yang/schema.lua]: please make this one line

File: src/lib/yang/schema.lua [hunk 2]
```diff
@@ -673,7 +681,14 @@ function resolve(schema, features)
       end
       return ret
    end
-   function visit_type(node, env)
+   local function resolve_bases(bases, env)
```
Comment (wingo) [inline src/lib/yang/schema.lua]: Please add a comment: "Resolve argument of "base" statements to identity fqid and collect in a list."

File: src/lib/yang/schema.lua [hunk 3]
```diff
@@ -884,6 +908,25 @@ function load_schema_by_name(name, revision)
...
 
+function lookup_identity (fqid)
+   local schema_name, id = fqid:match("^([^:]*):(.*)$")
+   local schema, env = load_schema_by_name(schema_name)
+   local id_thunk = env.identities[id]
+   if not id_thunk then
+      error('no identity '..id..' in module '..schema_name)
+   end
+   return id_thunk() -- Force the lazy lookup.
+end
```
Comment (wingo) [inline src/lib/yang/schema.lua:918]: Add `lookup_identity = util.memoize(lookup_identity)` please.

File: src/lib/yang/schema.lua [hunk 4]
```diff
@@ -884,6 +908,25 @@ function load_schema_by_name(name, revision)
...
+
+function identity_is_instance_of (identity, fqid)
+   for _, base in ipairs(identity.bases) do
+      if base == fqid then return true end
+      local base_id = lookup_identity(base)
+      if identity_is_instance_of(base_id, fqid) then return true end
+   end
+   return false
+end
```
Comment (wingo) [inline src/lib/yang/schema.lua:928]: Likewise please memoize this function.

File: src/lib/yang/value.lua [hunk 5]
```diff
@@ -79,7 +79,15 @@ function types.decimal64.tostring(val)
 end
 
 types.empty = unimplemented('empty')
-types.identityref = unimplemented('identityref')
+
+types.identityref = {}
+function types.identityref.parse(str, what)
+   return assert(str, 'missing value for '..what)
```
Comment (wingo) [inline src/lib/yang/value.lua]: We should add a comment that these references are expanded in the validation phase.

**Review by wingo — CHANGES_REQUESTED** [review]
Looks good; would you mind making the indicated changes?  Thanks!

File: src/lib/yang/schema.lua (hunk 3 above)
Comment (dpino) [reply src/lib/yang/schema.lua:918]: I don't know why this comment is not marked as outdated. The change was addressed in the amend patch.

File: src/lib/yang/schema.lua (hunk 4 above)
Comment (dpino) [reply src/lib/yang/schema.lua:928]: Likewise this case.

**Review by wingo — APPROVED** [review]
thanks!

**Pull request merged**